File: PrintDistinctPKs/LoadMaxValues/java/LoadMaxValues.java
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.*;
import software.amazon.awssdk.core.SdkBytes;

import java.math.BigDecimal;
import java.util.Arrays;

public class LoadMaxValues {
    private static final String S_TABLE_NAME = "max-str-sk-test-java";
    private static final String N_TABLE_NAME = "max-num-sk-test-java";
    private static final String B_TABLE_NAME = "max-bin-sk-test-java";

    private static final String MAX_SORT_KEY_VALUE_S = String.valueOf(new char[256]).replace("\0", "\uDBFF\uDFFF");
    private static final BigDecimal MAX_SORT_KEY_VALUE_N = new BigDecimal("9.9999999999999999999999999999999999999E+125");
    private static final SdkBytes MAX_SORT_KEY_VALUE_B = SdkBytes.fromByteArray(new byte[1024]);

    private static DynamoDbClient dynamoDb;

    public static void main(String[] args) {
        String region = args[0];
        dynamoDb = DynamoDbClient.builder().region(software.amazon.awssdk.regions.Region.of(region)).build();

        createTable(S_TABLE_NAME, ScalarAttributeType.S);
        createTable(N_TABLE_NAME, ScalarAttributeType.N);
        createTable(B_TABLE_NAME, ScalarAttributeType.B);
        waitForTablesReady();
        insertItem(S_TABLE_NAME, MAX_SORT_KEY_VALUE_S);
        insertItem(N_TABLE_NAME, MAX_SORT_KEY_VALUE_N);
        insertItem(B_TABLE_NAME, MAX_SORT_KEY_VALUE_B);
    }

    private static void createTable(String tableName, ScalarAttributeType skType) {
        CreateTableRequest createTableRequest = CreateTableRequest.builder()
                .tableName(tableName)
                .keySchema(Arrays.asList(KeySchemaElement.builder().attributeName("pk").keyType(KeyType.HASH).build(),
                        KeySchemaElement.builder().attributeName("sk").keyType(KeyType.RANGE).build()))
                .attributeDefinitions(Arrays.asList(AttributeDefinition.builder().attributeName("pk").attributeType(ScalarAttributeType.S).build(),
                        AttributeDefinition.builder().attributeName("sk").attributeType(skType).build()))
                .billingMode(BillingMode.PAY_PER_REQUEST)
                .build();

        try {
            System.out.printf("Creating Table %s%n", tableName);
            dynamoDb.createTable(createTableRequest);
            System.out.printf("Table %s created successfully%n", tableName);
        } catch (Exception error) {
            System.out.println("Error creating table: " + error);
        }
    }

    private static void waitForTablesReady() {
        for (String tableName : Arrays.asList(S_TABLE_NAME, N_TABLE_NAME, B_TABLE_NAME)) {
            System.out.printf("Waiting for Table %s to be ready%n", tableName);
            while (true) {
                DescribeTableResponse table = dynamoDb.describeTable(DescribeTableRequest.builder().tableName(tableName).build());
                if (table.table().tableStatus().equals(TableStatus.ACTIVE)) {
                    System.out.printf("Table %s is ready%n", tableName);
                    break;
                } else {
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }

    private static void insertItem(String tableName, Object skValue) {
        PutItemRequest putItemRequest = PutItemRequest.builder()
                .tableName(tableName)
                .item(Map.of(
                        "pk", AttributeValue.builder().s("sample-pk-value").build(),
                        "sk", skValue instanceof String ? AttributeValue.builder().s((String) skValue).build()
                                : skValue

File: PrintDistinctPKs/LoadMaxValues/python/load_max_values.py
import boto3
import argparse
import decimal
import boto3.dynamodb.types
import time

S_TABLE_NAME = "max-str-sk-test-python"
N_TABLE_NAME = "max-num-sk-test-python"
B_TABLE_NAME = "max-bin-sk-test-python"
MAX_SORT_KEY_VALUE_S = str(256 * chr(0x10FFFF))
MAX_SORT_KEY_VALUE_N = decimal.Decimal('9.9999999999999999999999999999999999999E+125')
MAX_SORT_KEY_VALUE_B = boto3.dynamodb.types.Binary(b'\xFF' * 1024)

dynamodb = None

def create_table(tableName, skType):
    global dynamodb

    create_table_params = {
        "TableName": tableName,
        "KeySchema": [
            {"AttributeName": "pk", "KeyType": "HASH"},
            {"AttributeName": "sk", "KeyType": "RANGE"},
        ],
        "AttributeDefinitions": [
            {"AttributeName": "pk", "AttributeType": "S"},
            {"AttributeName": "sk", "AttributeType": skType},
        ],
        "BillingMode": "PAY_PER_REQUEST",
    }

    try:
        print(f'Creating Table {tableName}')
        dynamodb.create_table(**create_table_params)
        print(f'Table {tableName}created successfully')

    except Exception as error:
        print("Error creating table:", error)

def wait_for_tables_ready():
    global dynamodb

    for tableName in [S_TABLE_NAME, N_TABLE_NAME, B_TABLE_NAME]:
        print(f'Waiting for Table {tableName} to be ready')
        while True:
            table = dynamodb.Table(tableName)
            if table.table_status == "ACTIVE":
                print(f'Table {tableName} is ready')
                break
            else:
                time.sleep(1)

    
def insert_item(tableName, skType, skValue):
    global dynamodb

    table = dynamodb.Table(tableName)

    largest_number = decimal.Decimal("9.9999999999999999999999999999999999999e+125")

    put_item_params = {
        "pk": "sample-pk-value",
        "sk": skValue,
    }

    try:
        table.put_item(Item=put_item_params)
        print(f'Item inserted successfully into table "{tableName}"')
    except Exception as error:
        print("Error inserting item:", error)

def main(region):
    global dynamodb

    dynamodb = boto3.resource('dynamodb', region_name=region)
    create_table(S_TABLE_NAME, 'S')
    create_table(N_TABLE_NAME, 'N')
    create_table(B_TABLE_NAME, 'B')
    wait_for_tables_ready()
    insert_item(S_TABLE_NAME, 'S', MAX_SORT_KEY_VALUE_S)
    insert_item(N_TABLE_NAME, 'N', MAX_SORT_KEY_VALUE_N)
    insert_item(B_TABLE_NAME, 'B', MAX_SORT_KEY_VALUE_B)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--region", required=True, help="AWS region for the DynamoDB table")
    args = parser.parse_args()

    main(args.region)

